handout/di/angular2/avoiding_collisions_opaque_token.md

```diff
@@ -0,0 +1,98 @@
+# Avoiding Injection Collisions: OpaqueToken
+
+Since Angular allows the use of tokens as identifiers to its dependency injection system, one of the potential issues is using the same token to represent different entities. If, for example, the string `'token'` is used to inject an entity, it's possible that something totally unrelated also uses `'token'` to inject a different entity. When it comes time for Angular to resolve one of these entities, it might be resolving the wrong one. This behaviour might happen rarely, or are easy to resolve when it happens within a small team, but when it comes to multiple teams working separately on the same codebase or integration of 3rd party modules, the cost of these collisions goes up.
+
+Consider this example where the main app is a consumer of two modules: one that provides an email service and another that provides a logging service.
+
+_app/email/email.service.ts_
+```typescript
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class EmailService {
+  constructor(@Inject(apiConfig) public apiConfig) { }
+}
+```
+
+_app/email/email.module.ts_
+```typescript
+@NgModule({
+  providers: [ EmailService ],
+})
+export class EmailModule { }
+```
+
+The email service has an associated api that is trying to inject configuration settings identified by the string `'api-config'`.
```

Comment:
I think we need to explain a little bit more the problem. Below are my proposed changes for this section.

---

The email service requires a configuration object identified by the token `api-service` that will be provided by the DI system. Because we want this module to be used in other applications, and the value of the configuration object depends on the application characteristics, we should provide the value of the token `api-service` in `AppModule` where the `EmailModule` is imported.

_app/app.module.ts_

```ts
import { apiConfig } from './email/index';

@NgModule({
  ...
  providers: [
    { provide: apiConfig, useValue: { apiKey: 'email-config', context: 'registration' } },
  ]
})
export class AppModule {}
```

At the same time, another team is working on a reusable `LoggerModule` that also requires a configuration object. Not surprisingly, they decided to use the same token for their configuration object, `api-config`.